File: apps/cli/src/terminal-command-catalog.ts
import type {
  TerminalCommandCatalog,
  TerminalCommandCatalogExecutionTarget,
  TerminalCommandCatalogNode,
  TerminalCommandCatalogSurface,
} from "@arroba/kernel-client/kernel-types"
import type { CommandNode } from "./command-center-tree-projection.js"

export type TerminalCommandCatalogProjectionOptions = {
  surface?: TerminalCommandCatalogSurface
  executionTargets?: readonly TerminalCommandCatalogExecutionTarget[]
  includeNode?: (node: TerminalCommandCatalogNode) => boolean
}

export function commandTreeFromTerminalCommandCatalog(
  catalog: TerminalCommandCatalog | null | undefined,
  options: TerminalCommandCatalogProjectionOptions = {},
): CommandNode[] {
  return commandNodesFromCatalogNodes(catalog?.nodes ?? [], options)
}

function commandNodesFromCatalogNodes(
  nodes: readonly TerminalCommandCatalogNode[],
  options: TerminalCommandCatalogProjectionOptions,
): CommandNode[] {
  return nodes.flatMap((node) => {
    if (options.includeNode && !options.includeNode(node)) {
      return []
    }
    const children = commandNodesFromCatalogNodes(node.children ?? [], options)
    if (node.kind === "group" && (node.children?.length ?? 0) > 0 && children.length === 0) {
      return []
    }
    const surfaceMatches = !options.surface || node.surfaces.includes(options.surface)
    const targetMatches = !options.executionTargets || options.executionTargets.includes(node.execution_target)
    if ((!surfaceMatches || !targetMatches) && children.length === 0) {
      return []
    }
    return [commandNodeFromCatalogNode(node, children)]
  })
}

function commandNodeFromCatalogNode(
  node: TerminalCommandCatalogNode,
  children: readonly CommandNode[],
): CommandNode {
  return {
    id: node.id,
    label: node.label,
    description: node.description,
    value: node.value,
    kind: node.kind,
    executionTarget: node.execution_target,
    surfaces: node.surfaces,
    ...(node.intents?.length ? { intents: node.intents } : {}),
    ...(node.examples?.length ? { examples: node.examples } : {}),
    ...(node.dynamic_source != null ? { dynamicSource: node.dynamic_source } : {}),
    ...(node.search_aliases?.length ? { searchAliases: node.search_aliases } : {}),
    ...(children.length ? { children } : {}),
  }
}

File: apps/kernel/src/local/api/types/terminal_command_catalog.rs
use super::*;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GetTerminalCommandCatalogRequest;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GetTerminalOperationRegistryRequest;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TerminalOperationContract {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    pub description: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub search_aliases: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub intents: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub required_context: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub required_targets: Vec<String>,
    pub input_schema: serde_json::Value,
    pub result_kind: String,
    pub mutation: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub expected_projections: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub supported_surfaces: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub examples: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub parity_variants: Vec<String>,
    pub presentation_only: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TerminalOperationRegistry {
    pub revision: String,
    pub operations: Vec<TerminalOperationContract>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TerminalCommandCatalogNodeKind {
    Group,
    Command,
    PromptPrefix,
    Dynamic,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TerminalCommandCatalogExecutionTarget {
    Kernel,
    TerminalLocal,
    PromptPrefix,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TerminalCommandCatalogSurface {
    Session,
    WaitingRoom,
    WorkflowScreen,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TerminalCommandCatalogNode {
    pub id: String,
    pub label: String,
    pub description: String,
    pub value: String,
    pub kind: TerminalCommandCatalogNodeKind,
    pub execution_target: TerminalCommandCatalogExecutionTarget,
    pub surfaces: Vec<TerminalCommandCatalogSurface>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub search_aliases: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub intents: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub examples: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub dynamic_source: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub children: Vec<TerminalCommandCatalogNode>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TerminalCommandCatalog {
    pub revision: String,
    pub nodes: Vec<TerminalCommandCatalogNode>,
}
